### PR TITLE
factor out <langutil>, a new [static] library that is shared library

### PR DESCRIPTION
**Pull request by christianparpart**
Factor out <langutil>, a new [static] library that is shared between libsolidity (solidity language frontend) and future libyul (YUL language frontend)

The motivation behind this is to eliminate circular dependencies in our build system and end up also having a standalone libyul static library (PR following this one).

## CHANGES (AND THOUGHTS)

* moved general purpose language frontend files out of libsolidity/
* **also** moved Scanner and Token, as they're both shared between Solidity and Yul (for now), can be generalized (maybe later, not now)
* introduced new namespace `langutil` and put all underneath it, adapting the users of them
* also added `EVMVersion` out (as it is shared); however, that one should have a namespace such as `eth`  or `evm` as it's actually about the EVM here (could be more appropriate in another folder, but I didn't wanna overdo it here).
* Extracted CharSource into its own header/impl file, with the future vision of potentially having one lexer/token API per language (solidity, yul,) that could share this API in order allow transition from one recognizer to another (forth and back).

### CODE REVIEW

**Comment by ekpyron** [comment]
It can of course be changed anytime, but does anybody have a better name than ``libsolcommon`` or do you all agree that it's fine? I mean - the idea is to separate things away from solidity, so if someone has a better name we might not want to have ``sol`` in it (not sure what else to call it though).

**Comment by christianparpart** [comment]
> It can of course be changed anytime, but does anybody have a better name than `libsolcommon` or do you all agree that it's fine? I mean - the idea is to separate things away from solidity, so if someone has a better name we might not want to have `sol` in it (not sure what else to call it though).

ErrorReporter, SourceLocation and SecondarySourceLocation are surely compiler frontend utility classes,  Even ParserBase, if it wouldn't depend on Solidity Token type (can be mate template) and Scanner (can be made more generic).

**Comment by christianparpart** [comment]
wrt. the naming of `libsolcommon`, I'm much more in favor of `langutil` (language utility) currently. it is containing a lot of non-solidity-related language frontend helper APIs, and yeah, plus Scanner and Token. Except those two, everything else could be used by any language frontend (IMHO)

**Review by chriseth — DISMISSED** [review]
Please rename to `langutil` and move all `langutil` includes below the `libsolidity` includes.

**Comment by christianparpart** [comment]
@chriseth: I actually wanted to change the name to that one, too.

With regards to the directory move I am not entirely convinced (or clear) yet.
Do you mean just for include search path, or moving also the cpp files? I fear people may get confused that langutil "belongs" to libsolidity (it's rather libsolidity depends on langutil, and links against it). The only thing in common (for now) is Scanner/Token. Maybe I am just getting a little confused though.

**Comment by ekpyron** [comment]
@christianparpart I think the idea was just to correct the order of includes, so that ``<langutils/*>`` includes come after (i.e. "below") ``<libsolidity/*>`` includes, since we order includes from "specific" to "general". (Am I right @chriseth?)

**Comment by christianparpart** [comment]
yep. you may be right @ekpyron . :-)

**Comment by chriseth** [comment]
Exactly, @ekpyron

File: CMakeLists.txt [hunk 1]
```diff
@@ -44,6 +44,7 @@ include(EthOptions)
 configure_project(TESTS)
 
 add_subdirectory(libdevcore)
+add_subdirectory(langutil)
```
Comment (chriseth) [inline CMakeLists.txt:47]: Perhaps `liblangutil`?

File: langutil/CharStream.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <tuple>
+
+namespace langutil {
```
Comment (chriseth) [inline langutil/CharStream.h]: `{` to line of its own

File: langutil/Exceptions.h [hunk 3]
```diff
@@ -24,33 +24,33 @@
 
 #include <string>
```
Comment (chriseth) [inline langutil/Exceptions.h]: should move down

File: langutil/Exceptions.h [hunk 4]
```diff
@@ -77,7 +77,7 @@ class Error: virtual public Exception
 	/// helper functions
 	static Error const* containsErrorOfType(ErrorList const& _list, Error::Type _type)
 	{
-		for (auto e: _list)
+		for (auto e : _list)
```
Comment (chriseth) [inline langutil/Exceptions.h]:
```suggestion
		for (auto const& e: _list)
```

File: langutil/Exceptions.h [hunk 5]
```diff
@@ -86,7 +86,7 @@ class Error: virtual public Exception
 	}
 	static bool containsOnlyWarnings(ErrorList const& _list)
 	{
-		for (auto e: _list)
+		for (auto e : _list)
```
Comment (chriseth) [inline langutil/Exceptions.h]:
```suggestion
		for (auto const& e: _list)
```

File: libsolidity/CMakeLists.txt [hunk 6]
```diff
@@ -26,7 +26,7 @@ if (NOT (${Z3_FOUND} OR ${CVC4_FOUND}))
 endif()
 
 add_library(solidity ${sources} ${headers})
-target_link_libraries(solidity PUBLIC evmasm devcore ${Boost_FILESYSTEM_LIBRARY} ${Boost_SYSTEM_LIBRARY})
+target_link_libraries(solidity PUBLIC solcommon evmasm devcore ${Boost_FILESYSTEM_LIBRARY} ${Boost_SYSTEM_LIBRARY})
```
Comment (axic) [inline libsolidity/CMakeLists.txt:29]: Isn't this supposed to be `langutil`?

File: libsolidity/parsing/Scanner.h [hunk 7]
```diff
@@ -1,249 +1,33 @@
...
-	GNU General Public License for more details.
+    solidity is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
 
-	You should have received a copy of the GNU General Public License
-	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
-
-	This file is derived from the file "scanner.h", which was part of the
```
Comment (chriseth) [inline libsolidity/parsing/Scanner.h:17]: Should not be removed.

**Comment by axic** [comment]
A lot of the history is not kept with not moving files. Keeping the history is important when trying to find reasons for bugs and when they were introduced. Even though Github doesn’t use `git log —follow`, it can be done on the commandline.

Also I feel that breaking this into 3 parts makes more sense:
- changing namespace
- moving files
- splitting files

File: libsolidity/parsing/Scanner.h (hunk 7 above)
Comment (chriseth) [reply libsolidity/parsing/Scanner.h:17]: Ah sorry, it was moved.

Comment (axic) [reply libsolidity/parsing/Scanner.h:17]: Well, copied :)

**Comment by chriseth** [comment]
`solc --license` currently shows
```
...
The libsolidity/parsing/{scanner,token}.{h,cpp} files are dervied from
...
```
- please update these paths.

**Pull request closed without merging**